File: packages/core/src/kernel.ts
/**
 * Kernel evaluation per ARKY-KERNEL-v1 §5.
 *
 * Given a Commitment and a set of TIM receipts: bind each MeasureSpec symbol
 * from matching TIM evidence, evaluate its assertion (tri-valued), resolve the
 * ConsequenceSpec array, and produce a Decision (APPROVED / REJECTED /
 * INDETERMINATE) listing authorized verbs.
 */

import { evaluateAssertion, type TriState, type Symbols } from './assert.ts';
import type { Tim } from './tim.ts';

/** Core verbs registered in ARKY-REGISTRIES-v1 (v1). */
export const REGISTERED_VERBS = new Set([
  'arky:verb/pay@v1',
  'arky:verb/refund@v1',
  'arky:verb/slash@v1',
  'arky:verb/revoke@v1',
  'arky:verb/upgrade@v1',
  'arky:verb/signal@v1',
  'arky:verb/control@v1',
]);

export interface MeasureSpec {
  name: string;
  from?: string;
  assert: string;
  window?: { start?: string; end?: string; max_age?: string };
  profile?: string;
  require?: { min_witnesses?: number; device_class?: string[]; code?: string[] };
}

export interface Verb {
  name: string;
  args: Record<string, unknown>;
}
export interface ConsequenceSpec {
  if: string;
  then: Verb[];
  limits?: unknown;
}

export interface Commitment {
  scope: string;
  actor: string;
  intent: { do: string; [k: string]: unknown };
  measure: MeasureSpec[];
  consequence: ConsequenceSpec[];
  cid?: string;
  sig?: string;
  [k: string]: unknown;
}

export interface AssertionResult {
  name: string;
  result: TriState;
  input_value?: number | string | boolean;
  unit?: string;
  inputs?: string[];
  error?: string;
}

export type DecisionStatus = 'APPROVED' | 'REJECTED' | 'INDETERMINATE';

export interface Decision {
  status: DecisionStatus;
  assertions: AssertionResult[];
  authorized: Verb[];
  errors: string[];
}

interface EvalOptions {
  /** evaluation time (ISO) for window max_age; defaults to now */
  time?: string;
}

/** Select TIMs matching a MeasureSpec and return the latest by Notary tuple. */
function selectLatest(spec: MeasureSpec, tims: Tim[], evalTime: string): Tim | undefined {
  let candidates = tims.slice();

  // require.min_witnesses
  if (spec.require?.min_witnesses !== undefined) {
    const min = spec.require.min_witnesses;
    candidates = candidates.filter((t) => (t.time.witnesses?.length ?? 0) >= min);
  }
  // require.device_class (match measurement.device)
  if (spec.require?.device_class?.length) {
    candidates = candidates.filter((t) =>
      spec.require!.device_class!.includes((t.measurement as any).device),
    );
  }
  // require.code
  if (spec.require?.code?.length) {
    candidates = candidates.filter((t) =>
      spec.require!.code!.includes((t.measurement as any).code),
    );
  }
  // window
  if (spec.window) {
    candidates = candidates.filter((t) => withinWindow(t.time.ts, spec.window!, evalTime));
  }
  if (candidates.length === 0) return undefined;

  // Notary ordering tuple: (ts ASC, lamport ASC, identity.id ASC, cid ASC); pick last.
  candidates.sort((a, b) => {
    if (a.time.ts !== b.time.ts) return a.time.ts < b.time.ts ? -1 : 1;
    const la = (a.time as any).ordering?.lamport ?? 0,
      lb = (b.time as any).ordering?.lamport ?? 0;
    if (la !== lb) return la - lb;
    if (a.identity.id !== b.identity.id) return a.identity.id < b.identity.id ? -1 : 1;
    return (a.cid ?? '') < (b.cid ?? '') ? -1 : a.cid === b.cid ? 0 : 1;
  });
  return candidates[candidates.length - 1];
}

function withinWindow(
  ts: string,
  w: { start?: string; end?: string; max_age?: string },
  evalTime: string,
): boolean {
  const t = Date.parse(ts);
  if (w.start && t < Date.parse(w.start)) return false;
  if (w.end && t >= Date.parse(w.end)) return false;
  if (w.max_age) {
    const ageMs = Date.parse(evalTime) - t;
    if (ageMs > parseIsoDurationMs(w.max_age)) return false;
  }
  return true;
}

/** Minimal ISO-8601 duration -> ms (supports the PnDTnHnMnS forms used in vectors). */
export function parseIsoDurationMs(d: string): number {
  const m = d.match(/^P(?:(\d+)D)?(?:T(?:(\d+)H)?(?:(\d+)M)?(?:(\d+(?:\.\d+)?)S)?)?$/);
  if (!m) return NaN;
  const [, dd, hh, mm, ss] = m;
  return (
    (Number(dd || 0) * 86400 + Number(hh || 0) * 3600 + Number(mm || 0) * 60 + Number(ss || 0)) *
    1000
  );
}

/**
 * Evaluate a commitment against TIM receipts. `tims` are the candidate evidence
 * (already verified by the caller). Returns a Decision.
 */
export function evaluateKernel(
  commitment: Commitment,
  tims: Tim[],
  opts: EvalOptions = {},
): Decision {
  const evalTime = opts.time ?? new Date().toISOString();
  const errors: string[] = [];
  const assertions: AssertionResult[] = [];
  const symbols: Symbols = {};

  // Structural validation (§5 step 1): a commitment missing required fields is
  // invalid and cannot be evaluated -> REJECTED with kernel.invalid_commitment.
  if (!Array.isArray(commitment.measure) || !Array.isArray(commitment.consequence)) {
    errors.push('kernel.invalid_commitment');
    return { status: 'REJECTED', assertions, authorized: [], errors };
  }

  // Static validation (§5 step 1 / §8.1): every verb referenced by any
  // ConsequenceSpec.then MUST be registered, independent of the outcome. An
  // unregistered verb makes the whole commitment invalid -> REJECTED.
  for (const cons of commitment.consequence) {
    for (const v of cons.then) {
      if (!REGISTERED_VERBS.has(v.name)) {
        errors.push('kernel.unknown_verb');
        return { status: 'REJECTED', assertions, authorized: [], errors };
      }
    }
  }

  // Bind each MeasureSpec symbol and evaluate its assertion.
  for (const spec of commitment.measure) {
    const tim = selectLatest(spec, tims, evalTime);
    const ar: AssertionResult = { name: spec.name, result: 'INDETERMINATE' };
    if (!tim) {
      ar.result = 'INDETERMINATE';
      ar.error = 'no matching receipts';
      assertions.push(ar);
      continue;
    }
    const value = (tim.measurement as any).value;
    const unit = (tim.measurement as any).unit;
    symbols[spec.name] = { value, unit };
    ar.input_value = value;
    if (unit !== undefined) ar.unit = unit;
    ar.inputs = tim.cid ? [tim.cid] : [];

    const res = evaluateAssertion(spec.assert, symbols);
    ar.result = res.result;
    if (res.error) ar.error = res.error;
    assertions.push(ar);
  }

  // Aggregate: APPROVED if all PASS; INDETERMINATE if any INDETERMINATE; else REJECTED.
  const anyIndet = assertions.some((a) => a.result === 'INDETERMINATE');
  const allPass = assertions.length > 0 && assertions.every((a) => a.result === 'PASS');
  const overall: TriState = anyIndet ? 'INDETERMINATE' : allPass ? 'PASS' : 'FAIL';

  if (overall === 'INDETERMINATE') {
    return { status: 'INDETERMINATE', assertions, authorized: [], errors };
  }

  // Resolve consequences in order; first matching `if` authorizes its `then`.
  // (All verbs were already verified as registered above.)
  let authorized: Verb[] = [];
  for (const cons of commitment.consequence) {
    if (matchesOutcome(cons.if, overall, assertions)) {
      authorized = cons.then;
      break;
    }
  }

  const status: DecisionStatus =
    overall === 'PASS' && authorized.length > 0 ? 'APPROVED' : 'REJECTED';
  return { status, assertions, authorized: status === 'APPROVED' ? authorized : [], errors };
}

/** Match a ConsequenceSpec.if (literal PASS/FAIL/INDETERMINATE) to the outcome. */
function matchesOutcome(
  ifClause: string,
  overall: TriState,
  _assertions: AssertionResult[],
): boolean {
  const c = ifClause.trim();
  if (c === 'PASS') return overall === 'PASS';
  if (c === 'FAIL') return overall === 'FAIL';
  if (c === 'INDETERMINATE') return overall === 'INDETERMINATE';
  // Expression outcomes (advanced) not needed for K1 vectors; default no-match.
  return false;
}

File: packages/core-rs/src/kernel.rs
//! Kernel evaluation per ARKY-KERNEL-v1 §5. Bind MeasureSpec symbols from TIM
//! evidence, evaluate assertions (tri-valued), resolve consequences, and emit a
//! Decision. Mirrors @arky/core (TS) kernel.ts.

use crate::assert::{SymVal, Symbols, TriState, evaluate_assertion};
use serde_json::Value;

/// Core verbs registered in ARKY-REGISTRIES-v1 (v1).
pub const REGISTERED_VERBS: &[&str] = &[
    "arky:verb/pay@v1",
    "arky:verb/refund@v1",
    "arky:verb/slash@v1",
    "arky:verb/revoke@v1",
    "arky:verb/upgrade@v1",
    "arky:verb/signal@v1",
    "arky:verb/control@v1",
];

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DecisionStatus {
    Approved,
    Rejected,
    Indeterminate,
}

impl DecisionStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            DecisionStatus::Approved => "APPROVED",
            DecisionStatus::Rejected => "REJECTED",
            DecisionStatus::Indeterminate => "INDETERMINATE",
        }
    }
}

#[derive(Debug, Clone)]
pub struct AssertionResult {
    pub name: String,
    pub result: TriState,
    pub input_value: Option<SymVal>,
    pub unit: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone)]
pub struct Decision {
    pub status: DecisionStatus,
    pub assertions: Vec<AssertionResult>,
    pub authorized: Vec<String>, // verb URNs
    pub errors: Vec<String>,
}

fn json_to_symval(v: &Value) -> Option<SymVal> {
    match v {
        Value::Number(n) => n.as_f64().map(SymVal::Num),
        Value::String(s) => Some(SymVal::Str(s.clone())),
        Value::Bool(b) => Some(SymVal::Bool(*b)),
        _ => None,
    }
}

/// Minimal ISO-8601 duration -> milliseconds (PnDTnHnMnS subset used in vectors).
pub fn parse_iso_duration_ms(d: &str) -> Option<i64> {
    // Very small parser: P[<n>D][T[<n>H][<n>M][<n>S]]
    let s = d.strip_prefix('P')?;
    let (days_part, time_part) = match s.split_once('T') {
        Some((d, t)) => (d, Some(t)),
        None => (s, None),
    };
    let mut total: f64 = 0.0;
    if !days_part.is_empty() {
        let n = days_part.strip_suffix('D')?;
        total += n.parse::<f64>().ok()? * 86400.0;
    }
    if let Some(t) = time_part {
        let mut rest = t;
        for (suffix, mult) in [('H', 3600.0), ('M', 60.0), ('S', 1.0)] {
            if let Some(idx) = rest.find(suffix) {
                let (num, after) = rest.split_at(idx);
                total += num.parse::<f64>().ok()? * mult;
                rest = &after[1..];
            }
        }
    }
    Some((total * 1000.0) as i64)
}

/// Parse an RFC3339 timestamp to epoch millis (UTC, the 'Z' form vectors use).
fn parse_rfc3339_ms(ts: &str) -> Option<i64> {
    // Format: YYYY-MM-DDTHH:MM:SSZ (no fractional/offset in the vectors).
    let bytes = ts.as_bytes();
    if bytes.len() < 20 || bytes[10] != b'T' {
        return None;
    }
    let num = |a: usize, b: usize| ts[a..b].parse::<i64>().ok();
    let (y, mo, d) = (num(0, 4)?, num(5, 7)?, num(8, 10)?);
    let (h, mi, s) = (num(11, 13)?, num(14, 16)?, num(17, 19)?);
    // days since epoch via a civil-from-days algorithm (Howard Hinnant's).
    let y = if mo <= 2 { y - 1 } else { y };
    let era = if y >= 0 { y } else { y - 399 } / 400;
    let yoe = y - era * 400;
    let doy = (153 * (if mo > 2 { mo - 3 } else { mo + 9 }) + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    let days = era * 146097 + doe - 719468;
    Some((days * 86400 + h * 3600 + mi * 60 + s) * 1000)
}

fn within_window(ts: &str, window: &Value, eval_time: &str) -> bool {
    let t = match parse_rfc3339_ms(ts) {
        Some(v) => v,
        None => return false,
    };
    if let Some(start) = window.get("start").and_then(|v| v.as_str())
        && let Some(st) = parse_rfc3339_ms(start)
        && t < st
    {
        return false;
    }
    if let Some(end) = window.get("end").and_then(|v| v.as_str())
        && let Some(en) = parse_rfc3339_ms(end)
        && t >= en
    {
        return false;
    }
    if let Some(max_age) = window.get("max_age").and_then(|v| v.as_str())
        && let (Some(et), Some(ms)) = (parse_rfc3339_ms(eval_time), parse_iso_duration_ms(max_age))
        && et - t > ms
    {
        return false;
    }
    true
}

/// Select the latest TIM matching a MeasureSpec (require/window + Notary tuple).
fn select_latest<'a>(spec: &Value, tims: &'a [Value], eval_time: &str) -> Option<&'a Value> {
    let mut cands: Vec<&Value> = tims.iter().collect();

    if let Some(req) = spec.get("require") {
        if let Some(minw) = req.get("min_witnesses").and_then(|v| v.as_u64()) {
            cands.retain(|t| {
                t.get("time")
                    .and_then(|x| x.get("witnesses"))
                    .and_then(|w| w.as_array())
                    .map(|a| a.len() as u64)
                    .unwrap_or(0)
                    >= minw
            });
        }
        if let Some(dc) = req.get("device_class").and_then(|v| v.as_array()) {
            let allowed: Vec<&str> = dc.iter().filter_map(|x| x.as_str()).collect();
            cands.retain(|t| {
                t.get("measurement")
                    .and_then(|m| m.get("device"))
                    .and_then(|d| d.as_str())
                    .map(|d| allowed.contains(&d))
                    .unwrap_or(false)
            });
        }
    }
    if let Some(window) = spec.get("window") {
        cands.retain(|t| {
            t.get("time")
                .and_then(|x| x.get("ts"))
                .and_then(|ts| ts.as_str())
                .map(|ts| within_window(ts, window, eval_time))
                .unwrap_or(false)
        });
    }
    if cands.is_empty() {
        return None;
    }
    // Notary tuple: (ts, lamport, identity.id, cid).
    cands.sort_by(|a, b| {
        let key = |t: &Value| {
            (
                t["time"]["ts"].as_str().unwrap_or("").to_string(),
                t["time"]["ordering"]["lamport"].as_i64().unwrap_or(0),
                t["identity"]["id"].as_str().unwrap_or("").to_string(),
                t["cid"].as_str().unwrap_or("").to_string(),
            )
        };
        key(a).cmp(&key(b))
    });
    cands.last().copied()
}

/// Evaluate a commitment against TIM receipts. `eval_time` is RFC3339 (UTC).
pub fn evaluate_kernel(commitment: &Value, tims: &[Value], eval_time: &str) -> Decision {
    let mut errors = Vec::new();
    let mut assertions = Vec::new();

    let measure = commitment.get("measure").and_then(|v| v.as_array());
    let consequence = commitment.get("consequence").and_then(|v| v.as_array());
    if measure.is_none() || consequence.is_none() {
        errors.push("kernel.invalid_commitment".into());
        return Decision {
            status: DecisionStatus::Rejected,
            assertions,
            authorized: vec![],
            errors,
        };
    }
    let (measure, consequence) = (measure.unwrap(), consequence.unwrap());

    // Static verb-registry validation (independent of outcome).
    for cons in consequence {
        if let Some(then) = cons.get("then").and_then(|v| v.as_array()) {
            for verb in then {
                let name = verb.get("name").and_then(|v| v.as_str()).unwrap_or("");
                if !REGISTERED_VERBS.contains(&name) {
                    errors.push("kernel.unknown_verb".into());
                    return Decision {
                        status: DecisionStatus::Rejected,
                        assertions,
                        authorized: vec![],
                        errors,
                    };
                }
            }
        }
    }

    let mut symbols = Symbols::new();
    for spec in measure {
        let name = spec
            .get("name")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let assert_expr = spec.get("assert").and_then(|v| v.as_str()).unwrap_or("");
        let tim = select_latest(spec, tims, eval_time);
        let mut ar = AssertionResult {
            name: name.clone(),
            result: TriState::Indeterminate,
            input_value: None,
            unit: None,
            error: None,
        };
        match tim {
            None => {
                ar.error = Some("no matching receipts".into());
            }
            Some(t) => {
                let value = &t["measurement"]["value"];
                if let Some(sv) = json_to_symval(value) {
                    symbols.insert(name.clone(), sv.clone());
                    ar.input_value = Some(sv);
                }
                ar.unit = t["measurement"]["unit"].as_str().map(String::from);
                let res = evaluate_assertion(assert_expr, &symbols);
                ar.result = res.result;
                ar.error = res.error;
            }
        }
        assertions.push(ar);
    }

    let any_indet = assertions
        .iter()
        .any(|a| a.result == TriState::Indeterminate);
    let all_pass = !assertions.is_empty() && assertions.iter().all(|a| a.result == TriState::Pass);
    let overall = if any_indet {
        TriState::Indeterminate
    } else if all_pass {
        TriState::Pass
    } else {
        TriState::Fail
    };

    if overall == TriState::Indeterminate {
        return Decision {
            status: DecisionStatus::Indeterminate,
            assertions,
            authorized: vec![],
            errors,
        };
    }

    // First matching consequence authorizes its verbs.
    let mut authorized: Vec<String> = vec![];
    for cons in consequence {
        let if_clause = cons.get("if").and_then(|v| v.as_str()).unwrap_or("").trim();
        let matches = match if_clause {
            "PASS" => overall == TriState::Pass,
            "FAIL" => overall == TriState::Fail,
            "INDETERMINATE" => overall == TriState::Indeterminate,
            _ => false,
        };
        if matches {
            if let Some(then) = cons.get("then").and_then(|v| v.as_array()) {
                authorized = then
                    .iter()
                    .filter_map(|v| v.get("name").and_then(|n| n.as_str()).map(String::from))
                    .collect();
            }
            break;
        }
    }

    let status = if overall == TriState::Pass && !authorized.is_empty() {
        DecisionStatus::Approved
    } else {
        DecisionStatus::Rejected
    };
    Decision {
        status,
        assertions,
        authorized: if status == DecisionStatus::Approved {
            authorized
        } else {
            vec![]
        },
        errors,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn duration_parsing() {
        assert_eq!(parse_iso_duration_ms("PT5M"), Some(300_000));
        assert_eq!(parse_iso_duration_ms("PT1H"), Some(3_600_000));
        assert_eq!(parse_iso_duration_ms("P2D"), Some(172_800_000));
    }

    #[test]
    fn rfc3339_parsing() {
        // 2025-10-15T12:05:00Z minus 2025-10-15T12:00:00Z == 5 min.
        let a = parse_rfc3339_ms("2025-10-15T12:00:00Z").unwrap();
        let b = parse_rfc3339_ms("2025-10-15T12:05:00Z").unwrap();
        assert_eq!(b - a, 300_000);
    }
}
